sigmastate/src/main/scala/sigmastate/Values.scala

```diff
@@ -1522,4 +1523,121 @@ object Values {
+          ContractTemplate(treeVersion, name, description, constTypes, constValues, parameters, expressionTree, null)
+        )
+      )
+
+    @inline final def isRightParsed: Boolean = expressionTree.isRight
+
+    private[sigmastate] var _bytes: Array[Byte] = templateBytes
+
+    /** Serialized bytes of this template. */
+    final def bytes: Array[Byte] = {
+      if (_bytes == null) {
+        _bytes = ContractTemplateSerializer.DefaultSerializer.serializeContractTemplate(this)
+      }
+      _bytes
+    }
+
+    def applyTemplate(paramValues: IndexedSeq[Constant[SType]]): ErgoTree = {
+      require(paramValues.length == parameters.length,
+        s"Expected ${parameters.length} number of parameters, got ${paramValues.length}")
+      require(isRightParsed, "expressionTree must be parsable to be able to apply template")
+
+      val nConsts = constTypes.size
+      val constants = safeNewArray[Constant[SType]](nConsts)
+      val paramIndices = this.parameters.map(p => p.placeholder).toSet
+      val positionToParamIndex = this.parameters.indices.map(i => (i -> this.parameters(i).placeholder)).toMap
+
+//      cfor(0)(_ < nConsts, _ + 1) { i =>
+//        if ()
+//      }
+
+      val constValues = this.constValues match {
+        case Some(values) => values
+        case None => safeNewArray[Option[SType#WrappedType]](constTypes.size)
+      }
+
+      cfor(0)(_ < paramValues.length, _ + 1) { i =>
+        val pos = parameters(i).placeholder
+//        constValues(pos) = paramValues(i)
+      }
+
+      // TODO: Check that value and param types match.
+
+      ErgoTree(
+        ErgoTree.ConstantSegregationHeader,
+        Array[Constant[SType]](),
+        this.expressionTree.right.get
+      )
+    }
+
+    /** The default equality of case class is overridden to exclude `complexity`. */
+    override def canEqual(that: Any): Boolean = that.isInstanceOf[ContractTemplate]
+
+    override def hashCode(): Int = Objects.hash(treeVersion, name, description, constTypes, constValues, parameters, expressionTree)
+
+    override def equals(obj: Any): Boolean = (this eq obj.asInstanceOf[AnyRef]) ||
+      ((obj.asInstanceOf[AnyRef] != null) && (obj match {
+        case other: ContractTemplate =>
+          other.treeVersion == treeVersion && other.name == name && other.description == description && other.constTypes == constTypes && other.constValues == constValues && other.parameters == parameters && other.expressionTree == expressionTree
+        case _ => false
+      }))
+  }
+
+  object ContractTemplate {
+    def apply(name: String,
+              description: String,
+              constTypes: IndexedSeq[SType],
+              constValues: Option[IndexedSeq[Option[SType#WrappedType]]],
+              parameters: IndexedSeq[Parameter],
+              expressionTree: SigmaPropValue): ContractTemplate = {
+      new ContractTemplate(None, name, description, constTypes, constValues, parameters, Right(expressionTree))
+    }
```

Comment:
This probably not necessary once you remove Either.

Reply:
Still good to keep if we want to allow creating a ContractTemplate with version as None.